A31-xds-routeaction-support.md

```diff
@@ -0,0 +1,226 @@
+    some point after the channel knows the RPC will no longer call the Load
+    Balancing Policy's Pick method and before the end of the RPC.  This is
+    necessary for cluster lifecycle management, which will be explained in
+    detail later.
+1.  Information which will be passed to the Load Balancing Policy's Pick method
+    by the channel.  E.g. in Go this could be done via the Context, and Java may
+    use CallOptions.  If an interceptor is used for (2) it is possible that the
+    interceptor may be capable of satisfying this requirement directly.
+1.  An RPC status; if not OK, the RPC is terminated with this status.  This is
+    needed in case the RPC cannot be routed.  Like (3), this could be satisfied
+    via an interceptor.
+
+Note that, for reasons described below, the Load Balancing Policy's Update
+method must be called and return before the Config Selector may be used by the
+channel.
+
+
+### Implementation for xDS RouteAction Support
+
+The "xds" resolver will return a Config Selector which applies the xDS matchers
+to the RPC (using the Method name and RPC Header data provided to it) and uses
+those inputs to perform the initial routing steps for the RPC based upon the RDS
+Routes.  This will enable it to set the timeouts and other settings
+appropriately in the MethodConfig according to the RouteActions.  Since the
+routing decisions are made here, it will also return the cluster name to the
+top-level Load Balancing Policy (described in detail below) in order to inform
+the policy which endpoint picker to use for the RPC.  Note that the Config
+Selector will also be assuming the responsibility of the current
+`weighted_target` Load Balancing Policy.  This means two things: 1. connectivity
+state is not considered when making this routing decision, and 2. once a cluster
+is chosen for an RPC, retry attempts will only use this cluster.  This matches
+Envoy behavior, and also supersedes the behavior specified in [gRFC
+A28](https://github.com/grpc/proposal/blob/master/A28-xds-traffic-splitting-and-routing.md).
+
+Because routing is happening in the Config Selector, the "xds_routing"
+Load Balancing Policy will no longer be needed.  Instead, the
+top-level Load Balancing Policy will be named
+"xds_cluster_manager_experimental", and will be a simple aggregator of
+child "cds" Load Balancing Policies, although the config format allows
+for other types of child Load Balancing Policies for future-proofing.
+The Load Balancing Config provided by the Resolver will be a [JSON
+representation](developers.google.com/protocol-buffers/docs/proto3#json)
+of the following proto message:
+
+```proto
+message XDSClusterManagerConfig {
+    message Child {
+      repeated LoadBalancingConfig child_policy = 1;
+    }
+    map<string, Child> children = 1;
+}
+```
+
+The "xds_cluster_manager" policy will maintain a map from cluster name (in each
+`cds_config` entry) to child Policy in order to route Picks.
```

Comment:
I'm not sure what the "in each cds_config entry" part is referring to.  I think the parenthetical can just be removed.

Reply:
I believe that is legacy from an older version of the design.